src/envoy/http/authn/authenticator_base.cc

```diff
@@ -36,6 +36,16 @@ AuthenticatorBase::~AuthenticatorBase() {}
 void AuthenticatorBase::done(bool success) const { done_callback_(success); }
 
 void AuthenticatorBase::validateX509(
+    const iaapi::MutualTls& mtls,
+    const AuthenticatorBase::MethodDoneCallback& done_callback) const {
+  if (mtls.allow_tls()) {
+    validateTls(mtls, done_callback);
```

Comment:
I still don't think splitting function is needed. Simply check the flag in the place where you need to get principal. Smt like

```cc
   MtlsAuthentication mtls_authn(filter_context_.connection());
   if (!mtls_authn.IsTLS()) {
     done_callback(nullptr, false);
     return;	
   }
 	 
   Payload payload;
   const auto hasCert = mtls_authn.GetSourceUser(payload.mutable_x509()->mutable_user()));
   if (!hasCert && !mtls.allow_tls()) {
      done_callback(nullptr, false);
  }
```

Also, I find this MtlsAuthentication class is over-killed, all those functions are single line. Please remove it and just put them inline.

Reply:
Separating the handling of mTLS and TLS makes the code easier to understand and less error prone. Will remove MtlsAuthentication class.

Reply:
It doesn't. They are not deviate enough to be treated differently. Split them early will cause more maintenance cost to keep them in sync.

Reply:
The first if can be just IsTLS() (the other one is covered by getsourceuser())

The rest is as my previous comment.

Reply:
As matter of fact, even the first if can be removed completely, if adding a check for cert available in getsourceuser.

Reply:
As matter of fact, even the first if can be removed completely, if adding a check for cert available in getsourceuser.

Reply:
Can you provide your proposed validateX509() implementation in code?

Reply:
Smt like this:

```cc
   if ( connection_ == nullptr || connection_->ssl() != nullptr)
     done_callback(nullptr, false);
     return;	
   }
 	 
   Payload payload;
   const auto hasCert = 
       connection_->ssl()->peerCertificatePresented() &&
       mtls_authn.GetSourceUser(payload.mutable_x509()->mutable_user()));

   if (!hasCert && !mtls.allow_tls()) {
      done_callback(nullptr, false);
  }
  done_callback(&payload, true);
```

Reply:
The above code has the following problems:
1. when connection_->ssl() is non-null, the above code calls done_callback(nullptr, false) and returns. A fix for this problem is to change "connection_->ssl() != nullptr" into "connection_->ssl() == nullptr".

2. GetSourceUser() can return true while the user returned is empty, which is not handled by this code. A fix for this problem is to change GetSourceUser() to return false when the user is empty.

3. when !hasCert && !mtls.allow_tls(), the above code will first call done_callback(nullptr, false) then call done_callback(&payload, true). To fix this problem, after done_callback(nullptr, false), the code should return.

I will fix the above problems and commit a new version.

Reply:
Done.